File: utils/codes/statistics.R
#' Actividad 2 - Verificacion cruzada en R del analisis estadistico.
#'
#' Lee el CSV generado por statistics.py, recalcula de forma independiente la
#' distribucion de frecuencias (Sturges), las medidas de tendencia central
#' (media, mediana, moda por clase modal interpolada) y de dispersion (rango,
#' varianza, desviacion estandar, coeficiente de variacion), y replica los
#' graficos basicos con la graficacion base, trazando la cuadricula en dos
#' pasadas para mantenerla detras de los datos.
#'
#' Escribe las imagenes en public/assets/images/figures/r/statistics/.

#' Rutas del proyecto (ejecutar desde la raiz del proyecto).
data_path <- "data/dataset/consumo_energia.csv"
figures_dir <- file.path("public", "assets", "images", "figures", "r", "statistics")
if (!dir.exists(figures_dir)) {
  dir.create(figures_dir, recursive = TRUE)
}

#' Carga de datos y vista rapida de su estructura.
df <- read.csv(data_path)
head(df)
str(df)

x <- df$consumo_kwh
n <- nrow(df)

#' 1. DISTRIBUCION DE FRECUENCIAS (regla de Sturges).
#'
#' Mismas k clases de igual amplitud que en Python; la tabla reporta fi, Fi,
#' hi (%) y Hi (%) para la verificacion cruzada.
k <- ceiling(1 + 3.322 * log10(n))
edges <- seq(min(x), max(x), length.out = k + 1)
fi <- hist(x, breaks = edges, plot = FALSE, include.lowest = TRUE)$counts
freq_table <- data.frame(
  lim_inferior = round(edges[-(k + 1)], 1),
  lim_superior = round(edges[-1], 1),
  marca_clase = round((edges[-(k + 1)] + edges[-1]) / 2, 1),
  fi = fi,
  Fi = cumsum(fi),
  hi_pct = round(fi / n * 100, 1),
  Hi_pct = round(cumsum(fi) / n * 100, 1)
)
print(freq_table)

#' Moda interpolada dentro de la clase modal: Mo = L + d1 / (d1 + d2) * w.
interpolated_mode <- function(values, bin_edges) {
  counts <- hist(values, breaks = bin_edges, plot = FALSE,
                 include.lowest = TRUE)$counts
  m <- which.max(counts)
  width <- bin_edges[m + 1] - bin_edges[m]
  d1 <- counts[m] - ifelse(m > 1, counts[m - 1], 0)
  d2 <- counts[m] - ifelse(m < length(counts), counts[m + 1], 0)
  if (d1 + d2 == 0) {
    return(bin_edges[m] + width / 2)
  }
  bin_edges[m] + d1 / (d1 + d2) * width
}

#' 2 y 3. TENDENCIA CENTRAL Y DISPERSION por sector y a nivel global.
#'
#' var() y sd() de R son muestrales (ddof = 1), igual que en pandas, por lo
#' que los valores deben coincidir digito a digito con central_tendency.csv y
#' dispersion.csv.
summary_stats <- function(values) {
  k_g <- ceiling(1 + 3.322 * log10(length(values)))
  edges_g <- seq(min(values), max(values), length.out = k_g + 1)
  c(
    n = length(values),
    media = round(mean(values), 1),
    mediana = round(median(values), 1),
    moda_interpolada = round(interpolated_mode(values, edges_g), 1),
    rango = round(max(values) - min(values), 1),
    varianza = round(var(values), 1),
    desv_std = round(sd(values), 1),
    cv_pct = round(sd(values) / mean(values) * 100, 1)
  )
}

groups <- c("Residencial", "Comercial", "Industrial")
stats_by_group <- t(sapply(groups, function(s) summary_stats(x[df$sector == s])))
stats_global <- summary_stats(x)
print(rbind(stats_by_group, Global = stats_global))
cat("Moda del sector (variable nominal):",
    names(which.max(table(df$sector))), "\n")

#' 4. GRAFICOS BASICOS (replicas de las figuras de Python).
#'
#' Principios aplicados: titulo informativo, ejes con unidades, cuadricula
#' sutil detras de los datos, color funcional y etiquetas de datos.

#' Histograma con clases de Sturges y medidas de tendencia central.
#'
#' La primera pasada traza un histograma sin relleno para fijar ejes y
#' escala, la cuadricula se inserta detras y la segunda pasada dibuja las
#' barras encima.
png(file.path(figures_dir, "hist_sturges_central_tendency.png"),
    width = 1950, height = 1140, res = 300, type = "cairo")
par(mar = c(4, 4, 3, 1))
hist(x, breaks = edges, border = NA, col = NA, include.lowest = TRUE,
     main = sprintf("Distribución de frecuencias del consumo (R, %d clases)", k),
     xlab = "Consumo (kWh/mes)", ylab = "Frecuencia absoluta (clientes)")
grid()
hist(x, breaks = edges, col = "#2c7fb8", border = "white",
     include.lowest = TRUE, add = TRUE)
global_mode <- interpolated_mode(x, edges)
abline(v = c(global_mode, median(x), mean(x)),
       col = "#d95f02", lty = c(3, 2, 1), lwd = 1.4)
legend("topright", cex = 0.75, col = "#d95f02", lty = c(3, 2, 1), lwd = 1.4,
       legend = c(sprintf("Moda = %.0f", global_mode),
                  sprintf("Mediana = %.0f", median(x)),
                  sprintf("Media = %.0f", mean(x))))
dev.off()

#' Poligono de frecuencias y ojiva en un mismo lienzo de dos paneles.
png(file.path(figures_dir, "freq_polygon_ogive.png"),
    width = 2700, height = 1020, res = 300, type = "cairo")
par(mfrow = c(1, 2), mar = c(4, 4, 3, 1))
poly_x <- c(edges[1], freq_table$marca_clase, edges[k + 1])
poly_y <- c(0, freq_table$fi, 0)
plot(poly_x, poly_y, type = "n", main = "Polígono de frecuencias (R)",
     xlab = "Consumo (kWh/mes)", ylab = "Frecuencia absoluta (clientes)")
grid()
lines(poly_x, poly_y, type = "o", pch = 19, col = "#2b8cbe")
plot(c(edges[1], freq_table$lim_superior), c(0, freq_table$Hi_pct),
     type = "n", ylim = c(0, 105), main = "Ojiva (R)",
     xlab = "Consumo (kWh/mes)", ylab = "Frecuencia acumulada (%)")
grid()
lines(c(edges[1], freq_table$lim_superior), c(0, freq_table$Hi_pct),
      type = "o", pch = 19, col = "#2b8cbe")
abline(h = 50, col = "#d95f02", lty = 2)
legend("bottomright", cex = 0.75, col = "#d95f02", lty = 2,
       legend = sprintf("50 %% -> Mediana ~ %.0f kWh", median(x)))
dev.off()

#' Barras de frecuencia absoluta por sector, ordenadas por magnitud.
sector_freq <- sort(table(df$sector))
png(file.path(figures_dir, "bar_freq_by_sector.png"),
    width = 1950, height = 960, res = 300, type = "cairo")
par(mar = c(4, 4, 3, 1))
bar_pos <- barplot(sector_freq, col = NA, border = NA,
                   ylim = c(0, max(sector_freq) * 1.2),
                   main = "Frecuencia de clientes por sector (R)",
                   xlab = "Sector", ylab = "Frecuencia absoluta (clientes)")
grid(nx = NA, ny = NULL)
barplot(sector_freq, col = c("#a6bddb", "#74a9cf", "#2b8cbe"),
        add = TRUE, axes = FALSE)
text(x = bar_pos, y = sector_freq + max(sector_freq) * 0.07,
     labels = sprintf("%d (%.0f%%)", sector_freq, sector_freq / n * 100),
     cex = 0.8)
dev.off()

#' Diagrama de caja con la media superpuesta, en dos pasadas.
sector_order <- c("Residencial", "Comercial", "Industrial")
df$sector <- factor(df$sector, levels = sector_order)
png(file.path(figures_dir, "boxplot_dispersion_by_sector.png"),
    width = 1950, height = 1140, res = 300, type = "cairo")
par(mar = c(4, 4, 3, 1))
boxplot(consumo_kwh ~ sector, data = df, border = NA,
        ylim = c(0, max(x) * 1.12),
        main = "Dispersión del consumo por sector (R)",
        xlab = "Sector", ylab = "Consumo (kWh/mes)")
grid(nx = NA, ny = NULL)
boxplot(consumo_kwh ~ sector, data = df,
        col = c("#a6bddb", "#74a9cf", "#2b8cbe"), add = TRUE, axes = FALSE)
group_means <- tapply(x, df$sector, mean)
group_sd <- tapply(x, df$sector, sd)
points(seq_along(sector_order), group_means, col = "#d95f02", pch = 19)
group_max <- tapply(x, df$sector, max)
text(seq_along(sector_order), group_max + max(x) * 0.05,
     labels = sprintf("sigma = %.0f", group_sd), cex = 0.75)
dev.off()

cat("OK - verificacion cruzada y figuras de R generadas\n")

File: utils/codes/statistics.py
"""Actividad 2 - Análisis estadístico de datos y creación de gráficos básicos.

Genera (o regenera con la misma semilla de la Actividad 1) el dataset de
consumo energético, construye la distribución de frecuencias por el método
de Sturges, calcula las medidas de tendencia central (media, mediana, moda)
y de dispersión (rango, varianza, desviación estándar, coeficiente de
variación) y produce los gráficos básicos con Matplotlib.

Rutas: el script se ubica en codes -> utils -> raíz del proyecto.
Escribe el CSV en ``data/dataset``, las tablas estadísticas en
``data/processed`` y las imágenes en
``public/assets/images/figures/python/statistics/``.
"""

from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

PROJECT_ROOT = Path(__file__).resolve().parents[2]
DATA_DIR = PROJECT_ROOT / "data" / "dataset"
PROCESSED_DIR = PROJECT_ROOT / "data" / "processed"
FIGURES_DIR = (
    PROJECT_ROOT / "public" / "assets" / "images" / "figures" / "python" / "statistics"
)
for d in (DATA_DIR, PROCESSED_DIR, FIGURES_DIR):
    d.mkdir(parents=True, exist_ok=True)

plt.rcParams.update({
    "figure.dpi": 150, "font.size": 10, "axes.titlesize": 11,
    "axes.titleweight": "bold", "axes.grid": True, "grid.alpha": 0.3,
    "axes.axisbelow": True,
})

"""Dataset: idéntico al de la Actividad 1 (semilla fija 42) para dar
continuidad al análisis sobre las mismas 120 observaciones."""
rng = np.random.default_rng(42)
n = 120
sectors = rng.choice(
    ["Residencial", "Comercial", "Industrial"], size=n, p=[0.5, 0.3, 0.2]
)
base = {"Residencial": 250, "Comercial": 900, "Industrial": 2500}
spread = {"Residencial": 60, "Comercial": 220, "Industrial": 600}
consumption = np.array([rng.normal(base[s], spread[s]) for s in sectors]).clip(50)
tariff = {"Residencial": 820, "Comercial": 710, "Industrial": 640}
cost = consumption * np.array([tariff[s] for s in sectors]) * rng.normal(1, 0.04, n) / 1000

df = pd.DataFrame({
    "cliente_id": [f"CL-{i:03d}" for i in range(1, n + 1)],
    "sector": sectors,
    "consumo_kwh": consumption.round(1),
    "costo_miles_cop": cost.round(1),
})
df.to_csv(DATA_DIR / "consumo_energia.csv", index=False)

"""1. DISTRIBUCIÓN DE FRECUENCIAS (regla de Sturges).

k = 1 + 3.322*log10(n) clases de igual amplitud sobre el consumo global.
La tabla incluye límites, marca de clase, frecuencias absolutas (fi),
acumuladas (Fi), relativas (hi %) y relativas acumuladas (Hi %).
"""
x = df["consumo_kwh"]
k = int(np.ceil(1 + 3.322 * np.log10(n)))
edges = np.linspace(x.min(), x.max(), k + 1)
fi, _ = np.histogram(x, bins=edges)
freq_table = pd.DataFrame({
    "lim_inferior": edges[:-1].round(1),
    "lim_superior": edges[1:].round(1),
    "marca_clase": ((edges[:-1] + edges[1:]) / 2).round(1),
    "fi": fi,
    "Fi": np.cumsum(fi),
    "hi_pct": (fi / n * 100).round(1),
    "Hi_pct": (np.cumsum(fi) / n * 100).round(1),
})
freq_table.to_csv(PROCESSED_DIR / "freq_table.csv", index=False)
print(f"Distribución de frecuencias ({k} clases, amplitud {edges[1]-edges[0]:.1f} kWh)")
print(freq_table.to_string(index=False))

"""2. MEDIDAS DE TENDENCIA CENTRAL.

La moda del consumo (variable continua) se reporta como clase modal e
interpolación: Mo = L + d1/(d1+d2) * w. La moda de la variable nominal
``sector`` es simplemente la categoría más frecuente.
"""


def interpolated_mode(values: pd.Series, bin_edges: np.ndarray) -> float:
    # Moda interpolada dentro de la clase modal (fórmula de agrupados).
    counts, _ = np.histogram(values, bins=bin_edges)
    m = int(np.argmax(counts))
    width = bin_edges[m + 1] - bin_edges[m]
    d1 = counts[m] - (counts[m - 1] if m > 0 else 0)
    d2 = counts[m] - (counts[m + 1] if m < len(counts) - 1 else 0)
    if d1 + d2 == 0:
        return float(bin_edges[m] + width / 2)
    return float(bin_edges[m] + d1 / (d1 + d2) * width)


central_rows = []
for label, g in [("Residencial", None), ("Comercial", None),
                 ("Industrial", None), ("Global", None)]:
    subset = x if label == "Global" else df.loc[df["sector"] == label, "consumo_kwh"]
    k_g = int(np.ceil(1 + 3.322 * np.log10(len(subset))))
    edges_g = np.linspace(subset.min(), subset.max(), k_g + 1)
    central_rows.append({
        "grupo": label,
        "n": len(subset),
        "media": round(subset.mean(), 1),
        "mediana": round(subset.median(), 1),
        "moda_interpolada": round(interpolated_mode(subset, edges_g), 1),
    })
central = pd.DataFrame(central_rows)
central.to_csv(PROCESSED_DIR / "central_tendency.csv", index=False)
print("\nMedidas de tendencia central del consumo (kWh)")
print(central.to_string(index=False))
print(f"Moda del sector (variable nominal): {df['sector'].mode()[0]}")

"""3. MEDIDAS DE DISPERSIÓN.

Varianza y desviación estándar muestrales (ddof=1), rango, coeficiente de
variación y rango intercuartílico, por sector y a nivel global.
"""
dispersion_rows = []
for label in ["Residencial", "Comercial", "Industrial", "Global"]:
    subset = x if label == "Global" else df.loc[df["sector"] == label, "consumo_kwh"]
    q1, q3 = subset.quantile([0.25, 0.75])
    dispersion_rows.append({
        "grupo": label,
        "rango": round(subset.max() - subset.min(), 1),
        "varianza": round(subset.var(ddof=1), 1),
        "desv_std": round(subset.std(ddof=1), 1),
        "cv_pct": round(subset.std(ddof=1) / subset.mean() * 100, 1),
        "q1": round(q1, 1),
        "q3": round(q3, 1),
        "iqr": round(q3 - q1, 1),
    })
dispersion = pd.DataFrame(dispersion_rows)
dispersion.to_csv(PROCESSED_DIR / "dispersion.csv", index=False)
print("\nMedidas de dispersión del consumo (kWh)")
print(dispersion.to_string(index=False))

"""4. GRÁFICOS BÁSICOS.

Principios aplicados (Actividad 1): título informativo, ejes con unidades,
eje de frecuencias desde cero, cuadrícula sutil detrás de los datos, color
funcional y etiquetas de datos.
"""

"""Histograma con las clases de Sturges y las tres medidas de tendencia
central superpuestas: evidencia la asimetría positiva del consumo global."""
global_mean = x.mean()
global_median = x.median()
global_mode = interpolated_mode(x, edges)
fig, ax = plt.subplots(figsize=(6.5, 3.8))
ax.hist(x, bins=edges, color="#2c7fb8", edgecolor="white")
for value, name, style in [
    (global_mode, f"Moda = {global_mode:,.0f}", ":"),
    (global_median, f"Mediana = {global_median:,.0f}", "--"),
    (global_mean, f"Media = {global_mean:,.0f}", "-"),
]:
    ax.axvline(value, color="#d95f02", linestyle=style, lw=1.4, label=name)
ax.set_title(f"Distribución de frecuencias del consumo ({k} clases de Sturges)")
ax.set_xlabel("Consumo (kWh/mes)")
ax.set_ylabel("Frecuencia absoluta (clientes)")
ax.legend(fontsize=8)
fig.tight_layout()
fig.savefig(FIGURES_DIR / "hist_sturges_central_tendency.png")
plt.close(fig)

"""Polígono de frecuencias y ojiva: dos vistas complementarias de la misma
tabla, la puntual (marcas de clase vs fi) y la acumulada (límites
superiores vs Hi %)."""
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(9.0, 3.4))
poly_x = np.concatenate(([edges[0]], freq_table["marca_clase"], [edges[-1]]))
poly_y = np.concatenate(([0], freq_table["fi"], [0]))
ax1.plot(poly_x, poly_y, marker="o", color="#2b8cbe")
ax1.set_title("Polígono de frecuencias")
ax1.set_xlabel("Consumo (kWh/mes)")
ax1.set_ylabel("Frecuencia absoluta (clientes)")
ax2.plot(np.concatenate(([edges[0]], freq_table["lim_superior"])),
         np.concatenate(([0], freq_table["Hi_pct"])),
         marker="o", color="#2b8cbe")
ax2.axhline(50, color="#d95f02", linestyle="--", lw=1,
            label=f"50 % -> Mediana ≈ {global_median:,.0f} kWh")
ax2.set_title("Ojiva (frecuencia acumulada)")
ax2.set_xlabel("Consumo (kWh/mes)")
ax2.set_ylabel("Frecuencia acumulada (%)")
ax2.set_ylim(0, 105)
ax2.legend(fontsize=8)
fig.tight_layout()
fig.savefig(FIGURES_DIR / "freq_polygon_ogive.png")
plt.close(fig)

"""Barras de frecuencia absoluta por sector: distribución de la variable
nominal, ordenada por magnitud y con etiquetas n (%)."""
sector_freq = df["sector"].value_counts().sort_values()
fig, ax = plt.subplots(figsize=(6.5, 3.2))
bars = ax.bar(sector_freq.index, sector_freq.values,
              color=["#a6bddb", "#74a9cf", "#2b8cbe"])
ax.bar_label(bars, labels=[f"{v} ({v/n:.0%})" for v in sector_freq.values], padding=3)
ax.set_ylim(0, sector_freq.max() * 1.18)
ax.set_title("Frecuencia de clientes por sector (moda nominal: "
             f"{df['sector'].mode()[0]})")
ax.set_xlabel("Sector")
ax.set_ylabel("Frecuencia absoluta (clientes)")
fig.tight_layout()
fig.savefig(FIGURES_DIR / "bar_freq_by_sector.png")
plt.close(fig)

"""Barras agrupadas media vs mediana por sector: la cercanía de ambas
dentro de cada grupo evidencia simetría local pese a la asimetría global."""
sector_order = ["Residencial", "Comercial", "Industrial"]
central_plot = central.set_index("grupo").loc[sector_order]
pos = np.arange(len(sector_order))
width = 0.38
fig, ax = plt.subplots(figsize=(6.5, 3.6))
b1 = ax.bar(pos - width / 2, central_plot["media"], width,
            label="Media", color="#2b8cbe")
b2 = ax.bar(pos + width / 2, central_plot["mediana"], width,
            label="Mediana", color="#a6bddb")
ax.bar_label(b1, fmt="%.0f", padding=2, fontsize=8)
ax.bar_label(b2, fmt="%.0f", padding=2, fontsize=8)
ax.set_xticks(pos, sector_order)
ax.set_ylim(0, central_plot["media"].max() * 1.18)
ax.set_title("Media y mediana del consumo por sector")
ax.set_xlabel("Sector")
ax.set_ylabel("Consumo (kWh/mes)")
ax.legend(fontsize=8)
fig.tight_layout()
fig.savefig(FIGURES_DIR / "bar_mean_median_by_sector.png")
plt.close(fig)

"""Diagrama de caja con la media superpuesta: lectura conjunta de posición
(mediana), dispersión (IQR y bigotes) y su relación con sigma."""
data = [df.loc[df["sector"] == s, "consumo_kwh"] for s in sector_order]
fig, ax = plt.subplots(figsize=(6.5, 3.8))
bp = ax.boxplot(data, tick_labels=sector_order, patch_artist=True,
                medianprops=dict(color="black"))
for patch, c in zip(bp["boxes"], ["#a6bddb", "#74a9cf", "#2b8cbe"]):
    patch.set_facecolor(c)
sigma = dispersion.set_index("grupo").loc[sector_order, "desv_std"]
means = central.set_index("grupo").loc[sector_order, "media"]
ax.scatter(pos + 1, means, color="#d95f02", zorder=3, s=30, label="Media")
for i, s in enumerate(sector_order):
    ax.annotate(f"σ = {sigma[s]:,.0f}", (i + 1, means[s]),
                xytext=(8, -4), textcoords="offset points", fontsize=8)
ax.set_title("Dispersión del consumo por sector (mediana, IQR, media y σ)")
ax.set_xlabel("Sector")
ax.set_ylabel("Consumo (kWh/mes)")
ax.legend(fontsize=8, loc="upper left")
fig.tight_layout()
fig.savefig(FIGURES_DIR / "boxplot_dispersion_by_sector.png")
plt.close(fig)

print("\nOK - tablas estadísticas y figuras de Python generadas")
